lib/rules/no-undefined.js

```diff
@@ -21,15 +21,76 @@ module.exports = {
 
     create(context) {
 
+        /**
+         * Checks whether the given reference is 'undefined' or not.
+         *
+         * @param {escope.Reference} reference - The reference to check.
+         * @returns {boolean} `true` if the reference is 'undefined'.
+         */
+        function isUndefinedReference(reference) {
+            const id = reference.identifier;
+
+            return id && id.name === "undefined";
+        }
+
+        /**
+         * Report an invalid "undefined" identifier node.
+         * @param {ASTNode} node The node to report.
+         * @returns {void}
+         */
+        function report(node) {
+            context.report({
+                node,
+                message: "Unexpected use of undefined."
+            });
+        }
+
+        /**
+         * Checks the given scope for references to `undefined` and reports
+         * all references found.
+         * @param {escope.Scope} scope The scope to check.
+         * @returns {void}
+         */
+        function checkScope(scope) {
+            const undefinedVar = scope.variables
+                ? scope.variables.filter(variable => variable.name === "undefined")[0]
```

Comment:
`scope` has ES6 `Map` object which is mapping variable names to each variable. It would be more efficient.

```js
const undefinedVar = scope.set.get("undefined")
if (!undefinedVar) {
    return
}
```

(I'm not sure why the name of the `Map` is `set`.)